packages/flutter/lib/src/services/restoration.dart

```diff
@@ -0,0 +1,888 @@
+  /// be [dispose]d and replaced with a new bucket.
+  ///
+  /// As an example, the [RestorationManager] calls this method on its root
+  /// bucket when it has been asked to restore a running application to a
+  /// different state. At that point, the data stored in the current bucket
+  /// hierarchy is invalid and will be replaced with a new hierarchy generated
+  /// from the restoration data describing the new state. To replace the current
+  /// bucket hierarchy, [decommission] is called on the root bucket to signal to
+  /// all owners of buckets in the hierarchy that their bucket has become
+  /// invalid. In response to the notification, bucket owners must [dispose]
+  /// their buckets and claim a new bucket from the newly created hierarchy. For
+  /// example, the owner of a bucket that was originally claimed from the
+  /// [RestorationManager.rootBucket] must dispose that bucket and claim a new
+  /// bucket from the new [RestorationManager.rootBucket]. Once the new bucket
+  /// is claimed, owners should restore their state according to the data stored
+  /// in the new bucket.
+  ///
+  /// A call to [decommission] must always be followed by a call to [dispose].
+  void decommission() {
+    _parent?._dropChild(this);
+    _parent = null;
+    _performDecommission();
+  }
+
+  void _performDecommission() {
+    _manager = null;
+    notifyListeners();
+    _visitChildren((RestorationBucket bucket) {
+      bucket._performDecommission();
+    });
+  }
+
+  // Get and store values.
+
+  /// Returns the value that is currently stored in the bucket under the
+  /// provided `id`.
+  ///
+  /// Returns null if nothing is stored under that id.
+  ///
+  /// See also:
+  ///
+  ///  * [put], which stores a value in the bucket.
+  ///  * [remove], which removes a value from the bucket.
+  ///  * [containsValue], which checks whether any value is stored under a given
+  ///    [RestorationId].
+  P get<P>(RestorationId id) {
+    assert(id != null);
+    return _rawValues[id.value] as P;
+  }
+
+  /// Stores the provided `value` under the provided `id` in the bucket.
+  ///
+  /// Any value that has previously been stored under that id is overwritten
+  /// with the new value. The provided `value` must be serializable with the
+  /// [StandardMessageCodec].
+  ///
+  /// Null values will be stored in the bucket as-is. To remove a value, use
+  /// [remove].
+  ///
+  /// See also:
+  ///
+  ///  * [get], which retrieves a stored value from the bucket.
+  ///  * [remove], which removes a value from the bucket.
+  ///  * [containsValue], which checks whether any value is stored under a given
+  ///    [RestorationId].
+  void put<P>(RestorationId id, P value) {
+    assert(id != null);
+    assert(debugIsSerializableForRestoration(value));
+    if (_rawValues[id.value] != value || !_rawValues.containsKey(id.value)) {
+      _rawValues[id.value] = value;
+      _manager?.scheduleSerialization();
```

Comment:
This seems prone to issues for the caller - if the manager is null, this data might not actually get stored.

I think at the least we should document this, but it might be a nicer API if we asserted the manager isn't null here instead and required callers to make sure the manager is initialized.

Alternatively, we should make this return a bool as to whether it actually scheduled serialization or not, so that a caller can think to check that and initialize the manager if it's not already.

Reply:
My thinking is that those are two separate responsibilities: When I get handed a bucket, I can store all the data I want in it. It is, however, not my concern to decide how/where the data is persisted. That's the responsibility of whoever owns the bucket and provided it to me. They need to make sure that it is hooked up to a manager.

Not requiring a manager also has the advantage that we can unhook a subtree. For example, when a widget is in the way of transitioning out in an animation, we no longer need to persist its state (in the state we've already recorded that it is gone). We can then just unhook the bucket.